Eliminate deadlock for recompress_chunk

mkindahl opened the pull request:
When executing `recompress_chunk` and a query at the same time, a
deadlock can be generated because the chunk relation and the chunk
index and the compressed and uncompressd chunks are locked in different
orders. In particular, when `recompress_chunk` is executing, it will
first decompress the chunk and as part of that lock the uncompressed
chunk index in AccessExclusive mode and when trying to compress the
chunk again it will try to lock the uncompressed chunk in
AccessExclusive as part of truncating it.

Note that `decompress_chunk` and `compress_chunk` lock the relations in
the same order and the issue arises because the procedures are combined
inth a single transaction.

To avoid the deadlock, this commit rewrites the `recompress_chunk` to
be a procedure and adds a commit between the decompression and
compression. Committing the transaction after the decompress will allow
reads and inserts to proceed by working on the uncompressed chunk, and
the compression part of the procedure will take the necessary locks in
strict order, thereby avoiding a deadlock.

In addition, the isolation test is rewritten so that instead of adding
a waitpoint in the PL/SQL function, we implement the isolation test by
taking a lock on the compressed table after the decompression.

Fixes #3846

mfundul commented on tsl/src/compression/compress_utils.c:
```diff
@@ -573,6 +574,10 @@ bool
+	/* To avoid a deadlock, we conservatively lock the uncompressed chunk when
+	 * doing a recompress. This will prevent reads from the uncompressed chunk
+	 * while recompress chunk is running. */
+	LockRelationOid(uncompressed_chunk_relid, AccessExclusiveLock);
```
What about unlocking after recompress is done? This should help with performance, but all the locks need to be released to avoid the deadlock.

mkindahl replied:
Relation is automatically unlocked when the transaction commits, which is when recompress is done.

mfundul replied:
Can't recompress be done as part of a multi-step transaction block?

mfundul replied:
Ok, I guess it always will take place at the end, regardless of the isolation mode, is what you're saying?

mkindahl replied:
Yes, but in that case the user has chosen to start a transaction and we should not commit it prematurely nor release the locks prematurely because we think it is a good idea. This can create consistency issues. The normal use case is to run it at top level, and in this case it would be released on completion of the statement.

gayyappan replied:
The problem with taking a lock early in the process is that anything that needs access to this chunk will be blocked. All queries on this hypertable or any operations on this hypertable are completely blocked and will  be waiting for this lock. Recompress could be a very long operation. The deadlock issue was due to the order in which the locks were acquired. Some of the locks can be completely eliminated by avoiding some steps during recompress as outlined in the support issue. Why are we not implementing that approach instead of taking more heavy handed locks?

mkindahl replied:
Removing the strong lock in favor of not re-building the index for the uncompressed chunk, as we discussed. This will avoid taking an `AccessExclusive` lock on the uncompressed chunk and eliminate the immediate problem. Other improvements are possible as well, but we should take these as a separate PR.

mkindahl commented:
Split out the test in a separate commit so that it is easy see that the fix actually solves the problem.

mfundul commented on tsl/test/isolation/specs/deadlock_recompress_chunk.spec.in:
```diff
@@ -0,0 +1,71 @@
+    SELECT recompress_chunk(show_chunks('hyper'));
+}
+		      
+# Since the locking order is different, we locking before reindex we should one one lock on the index and
```
"we should one one lock" typo.

mkindahl replied:
Fixed.

mfundul commented on tsl/src/compression/compression.c:
```diff
@@ -1106,15 +1106,16 @@ decompress_chunk(Oid in_table, Oid out_table)
 
-	/* Recreate all indexes on out rel, we already have an exclusive lock on it,
-	 * so the strong locks taken by reindex_relation shouldn't matter. */
+	if (reindex_out_table)
```
Wasn't the index rebuilt so that it's not inconsistent with the relation? What is the side-effect of not rebuilding the index here? Are we sure the index is not used later in the recompression path?

mkindahl replied:
> Wasn't the index rebuilt so that it's not inconsistent with the relation?

Not sure I follow.

Yes, if the index is not rebuild, it might be inconsistent with the relation, but that is not relevant. The point here is that an option is added so that the caller can opt out of rebuilding the index when it is known that it is not necessary to rebuild the index. For example, because it will not be inconsistent (because it is rebuilt before the caller decides to decompress) or because it will later be overwritten (as is the case here). It is entirely up to the caller.

A more relevant question would be if the index needs to be rebuilt or not in each of the cases where this function is called. And in this case we only skip rebuilding the index when doing a decompression as part of a `recompress_chunk`.

> What is the side-effect of not rebuilding the index here?

I'm not sure what side-effects you had in mind. Could you expand a on what you are concerned about here?

> Are we sure the index is not used later in the recompression path?

The index is overwritten as part of the compression (inside `compress_chunk`) by using `reindex_relation` in a similar way to how it is done here. Look in that function if you want to check if the index is used.

mfundul replied:
I think I am confused with the terminology, in_table, out_table etc. By trying to understand the code, it appears to me that `compress_chunk()` reindexes a different relation than `decompress_chunk()`.

mkindahl replied:
No, that is right, but there is still no reason to reindex the uncompressed chunk when decompressing since the chunk will be compressed after.

I am still not sure what failure scenario you had in mind.

mfundul replied:
I am looking into whether `compress_chunk()` uses the uncompressed chunk index, which was not re-indexed, in any way. I haven't found anything yet.

mfundul reviewed (changes requested):
The fix is flaky, the deadlock is not always resolved. The reason is that `compress_chunk()` is calling `truncate_relation()` on the uncompressed table which does a `reindex_relation()` that may race with the query and cause a deadlock again.

We could probably remove `reindex_relation()` altogether from `truncate_relation()` if we are sure it's never used. When we need uncompressed data again `decompress_chunk()` will make sure it's reindexed at that point in time.

mkindahl commented:
> The fix is flaky, the deadlock is not always resolved. The reason is that compress_chunk() is calling truncate_relation() on the uncompressed table which does a reindex_relation() that may race with the query and cause a deadlock again.

> We could probably remove reindex_relation() altogether from truncate_relation() if we are sure it's never used. When we need uncompressed data again decompress_chunk() will make sure it's reindexed at that point in time.

I do not think that is the case. Queries take the locks on the chunk first and then the chunk index, and in this case `truncate_relation()` uncompressed chunk is first locked, and then `reindex_relation()` will take the lock on the chunk index, so this does not seem to solve the problem (because locks on chunks and indexes are taken in the same order).

I manage to do some sort of reproduction locally, and get this locking table at the waitpoint (after decompression but before compression). As you can see, there is no indexes locked in exclusive mode there , but the compressed table is locked in `AccessExclusive` (from the code, because it is dropped at the end of the decompress), together with the toast table and toast table index.

```
422898|   12141|pg_locks                                                 |AccessShareLock         |t      
422899|   23345|_timescaledb_internal._hyper_141_300_chunk_time_plain    |AccessShareLock         |t      
422899|   23340|hyper_time_idx                                           |AccessShareLock         |t      
422899|   23338|time_plain                                               |AccessShareLock         |t      
422899|   23335|hyper                                                    |AccessShareLock         |t      
422899|   23346|_timescaledb_internal._hyper_141_300_chunk_hyper_time_idx|AccessShareLock         |t      
422899|   23354|_timescaledb_internal.compress_hyper_142_301_chunk       |AccessShareLock         |f      
422899|   23341|_timescaledb_internal._hyper_141_300_chunk               |AccessShareLock         |t      
422900|   23347|_timescaledb_internal._compressed_hypertable_142         |AccessShareLock         |t      
422900|   23335|hyper                                                    |AccessShareLock         |t      
422900|   23359|pg_toast.pg_toast_23354_index                            |AccessExclusiveLock     |t      
422900|   23354|_timescaledb_internal.compress_hyper_142_301_chunk       |AccessExclusiveLock     |t      
422900|   23341|_timescaledb_internal._hyper_141_300_chunk               |AccessShareLock         |t      
422900|   23341|_timescaledb_internal._hyper_141_300_chunk               |ShareUpdateExclusiveLock|t      
422900|   23341|_timescaledb_internal._hyper_141_300_chunk               |ExclusiveLock           |t      
422900|   23357|pg_toast.pg_toast_23354                                  |AccessExclusiveLock     |t      
422900|   23354|_timescaledb_internal.compress_hyper_142_301_chunk       |ExclusiveLock           |t      
422900|   16488|_timescaledb_catalog.chunk                               |RowExclusiveLock        |t      
422900|   16661|_timescaledb_catalog.hypertable_compression              |AccessShareLock         |t      
```

So, when I continue execution from this point, I get this deadlock graph. So now the problem is that the recompress process take the lock on the compressed chunk early (because it is dropping the compressed chunk), which then create a deadlock because the query take the lock late.

![locks](https://user-images.githubusercontent.com/1175692/143555194-b06a0e62-e7cf-4c95-b0d8-718f429e742e.png)

One way to solve this is to take a lock on the uncompressed chunk first (as originally proposed), but that would prevent reads from the uncompressed chunk while the recompress is running.

Another approach is to eliminate the need to take an exclusive lock on the compressed chunk, but the problem here is that if we do not truncate or drop the compressed table, we will get duplicate results. I was briefly considering if we could take a short-lived lock on the catalog table and replace it with an empty table that we did not keep locked, but that also contains a race condition since we do not follow a strict locking scheme in that case and release locks early and will take more locks later.

mfundul commented:
This looks like a cross-chunk deadlock (chunk 300 and 301). Wouldn't that mean the query needs to scan the chunks in the same order as the recompression process for no deadlock to exist?

mkindahl commented:
> This looks like a cross-chunk deadlock (chunk 300 and 301). Wouldn't that mean the query needs to scan the chunks in the same order as the recompression process for no deadlock to exist?

No, this is the compressed chunk for chunk 300. There is only one chunk in the table. Here is the contents of the chunk catalog:

```
    SELECT id, hypertable_id, schema_name, table_name, compressed_chunk_id
    FROM _timescaledb_catalog.chunk;

 id|hypertable_id|schema_name          |table_name                  |compressed_chunk_id
---+-------------+---------------------+----------------------------+-------------------
301|          142|_timescaledb_internal|compress_hyper_142_301_chunk|                   
300|          141|_timescaledb_internal|_hyper_141_300_chunk        |                301
(2 rows)
```

mkindahl commented:
There is actually one additional alternative and that is to run the `recompress_chunk` as two separate transactions. One to decompress, and then one to compress. This will release all locks after the decompression and allow queries to proceed, and then perform the compression part of the work in parallel with the queries. This would mean that a `recompress_chunk` is not transactional though.

mfundul commented:
One idea is to disallow lock upgrades altogether in our code. E.g. if one path takes `weak-lock`, and then it upgrades to `strong-lock` and blocks, then it holds `weak-lock` for no good reason, disallowing any cleanup from other contexts.

mfundul commented on tsl/src/compression/compress_utils.c:
```diff
@@ -334,13 +342,22 @@ decompress_chunk_impl(Oid uncompressed_hypertable_relid, Oid uncompressed_chunk_
-	ts_chunk_drop(compressed_chunk, DROP_RESTRICT, -1);
-	/* reenable autovacuum if necessary */
+
+	/* We only take an exclusive lock on the uncompressed chunk to prevent a
+	   deadlock if we are not going to reindex it. This is only when
+	   recompressing a chunk */
```
If this really is only when recompressing then it should be:
```
decompress_chunk_impl(Oid uncompressed_hypertable_relid, Oid uncompressed_chunk_relid,
					  bool if_compressed, bool recompressing_chunk)
...
	decompress_chunk(compressed_chunk->table_id,
					 uncompressed_chunk->table_id,
					 recompressing_chunk ? /* reindex_out_table */ false : true);
...
	if (recompressing_chunk)
		LockRelationOid(uncompressed_chunk->table_id, AccessExclusiveLock);
```
and keep the rest:
```
decompress_chunk(Oid in_table, Oid out_table, bool reindex_out_table)
```

mkindahl replied:
That seems to be in the wrong order. The recompress inside `decompress_chunk` will take a lock on the index, but the order should be: relation first, index after.

mfundul replied:
I'm not suggesting any change in order, I'm just renaming `reindex_out_table` to `recompressing_chunk` because it's very confusing as it currently is.

mfundul replied:
This way we don't need the comment either.

mkindahl commented:
Tested defining a procedure similar to this to handle the recompression works and eliminates the deadlock. Advisory locks added to create a debug waitpoint and behaves similar to the currently added debug waitpoint.
```
    CREATE PROCEDURE recompress_chunk_ng(regclass) AS $$
    BEGIN
      PERFORM decompress_chunk($1, if_compressed => true);
      COMMIT;
      PERFORM pg_advisory_lock(debug_waitpoint_id('recompress_after_decompress'));
      PERFORM pg_advisory_unlock(debug_waitpoint_id('recompress_after_decompress'));
      PERFORM compress_chunk($1);
    END $$
    LANGUAGE plpgsql;
```

Drawback of using a procedure is that it will not support set-returning functions such as `show_chunks`, so this will error out:
```
SELECT recompress_chunk(show_chunks('hyper'));
```

mfundul commented:
Maybe we could play with code like this:
https://github.com/postgres/postgres/blob/REL_13_5/src/backend/commands/vacuum.c#L627

To be able to retry, release locks held, and not block holding exclusive locks. The recompression logic could retry until query traffic goes down.

mkindahl commented:
> Maybe we could play with code like this:
https://github.com/postgres/postgres/blob/REL_13_5/src/backend/commands/vacuum.c#L627
> [...]
>To be able to retry, release locks held, and not block holding exclusive locks. The recompression logic could retry until query traffic goes down.

That has the problem that since the retry logic will stop retrying after a while, the user will have to restart the job quite often. It would basically make the `max_retries` parameter useless and if a job start failing for good reasons that it cannot recover from, it will keep retrying indefinitely. That does not sound like an improvement, it just moves the problem.

mfundul commented:
> > Maybe we could play with code like this:
> > https://github.com/postgres/postgres/blob/REL_13_5/src/backend/commands/vacuum.c#L627
> > [...]
> > To be able to retry, release locks held, and not block holding exclusive locks. The recompression logic could retry until query traffic goes down.
> 
> That has the problem that since the retry logic will stop retrying after a while, the user will have to restart the job quite often. It would basically make the `max_retries` parameter useless and if a job start failing for good reasons that it cannot recover from, it will keep retrying indefinitely. That does not sound like an improvement, it just moves the problem to a different place.

It doesn't have to retry indefinitely. This could also be configurable, e.g. `recompress_chunk()` could have a grace period that retries until a timeout and then it could take the heavy locks to finish the job. This could get rid of the majority of the deadlocks while also not blocking readers for certain workloads.

mkindahl commented:
Updated the pull request to change `recompress_chunk` into a procedure and changed test accordingly. Still in draft since I think the original C functions can be removed, but would like to verify that before proceeding.

mkindahl commented:
> > > Maybe we could play with code like this:
> > > https://github.com/postgres/postgres/blob/REL_13_5/src/backend/commands/vacuum.c#L627
> > > [...]
> > > To be able to retry, release locks held, and not block holding exclusive locks. The recompression logic could retry until query traffic goes down.
> > 
> > 
> > That has the problem that since the retry logic will stop retrying after a while, the user will have to restart the job quite often. It would basically make the `max_retries` parameter useless and if a job start failing for good reasons that it cannot recover from, it will keep retrying indefinitely. That does not sound like an improvement, it just moves the problem to a different place.
> 
> It doesn't have to retry indefinitely. This could also be configurable, e.g. `recompress_chunk()` could have a grace period that retries until a timeout and then it could take the heavy locks to finish the job. This could get rid of the majority of the deadlocks while also not blocking readers for certain workloads.

If the system is overloaded, a heavy lock to "finish the work" will just make a bad situation worse, so we should really avoid this.

gayyappan commented on sql/maintenance_utils.sql:
```diff
@@ -29,7 +29,40 @@ CREATE OR REPLACE FUNCTION decompress_chunk(
+  status INT;
+  chunk_name TEXT[];
+BEGIN
+    status := _timescaledb_internal.chunk_status(chunk);
```
Could we use SQL instead to get the status?
select ch.status INTO status
from _timescaledb_catalog.chunk ch, pg_class 
where oid = '_timescaledb_internal._hyper_1_1_chunk'::regclass and relname = ch.table_name;

mfundul replied:
Does this work for distributed hypertables?

mkindahl replied:
@gayyappan It is significantly easier and less error-prone to to maintain the `chunk_status` function than it is to maintain a fairly complex SQL query. In addition, it should probably be

```sql
SELECT ch.status INTO status
  FROM _timescaledb_catalog.chunk ch, pg_class cl, pg_namespace ns
 WHERE cl.oid = chunk AND relnamespace = ns.oid
   AND nspname = ch.schema_name  AND relname = ch.table_name AND NOT ch.dropped;
```

Which sort of proves my point that it is hard to maintain and will not adapt well to changes in the schema.

mfundul replied:
I was referring to Gayathri's comment. Just wondering if `select ch.status INTO status
from _timescaledb_catalog.chunk ch, pg_class
where oid = '_timescaledb_internal._hyper_1_1_chunk'::regclass and relname = ch.table_name;` would work for distributed hypertables.

gayyappan commented on tsl/src/chunk_api.c:
```diff
@@ -332,6 +332,17 @@ get_hypercube_from_slices(Jsonb *slices, const Hypertable *ht)
+ * Get the chunk status.
+ */
+Datum
+chunk_status(PG_FUNCTION_ARGS)
```
we can eliminate this function by querying chunk table using SQL in the procedure.

gayyappan replied:
Any metadata changes will need changes to views that use chunks. But either way is fine with me.

mfundul commented on src/cross_module_fn.c:
```diff
@@ -93,6 +92,7 @@ CROSSMODULE_WRAPPER(chunk_get_relstats);
 CROSSMODULE_WRAPPER(chunk_get_colstats);
 CROSSMODULE_WRAPPER(chunk_create_empty_table);
 CROSSMODULE_WRAPPER(chunk_create_replica_table);
+CROSSMODULE_WRAPPER(chunk_status);
```
Almost all code that gets catalog metadata uses the Apache license. `chunk_status()` only calls `ts_chunk_get_by_relid()` which is also an Apache function. I don't think `chunk_status()` should be a TSL function.

mkindahl replied:
That makes sense. Move the function.

mfundul commented on sql/chunk.sql:
```diff
@@ -21,6 +21,10 @@ CREATE OR REPLACE FUNCTION _timescaledb_internal.calculate_chunk_interval(
 ) RETURNS BIGINT AS '@MODULE_PATHNAME@', 'ts_calculate_chunk_interval' LANGUAGE C;
 
+-- Get the status of the chunk
+CREATE OR REPLACE FUNCTION _timescaledb_internal.chunk_status(REGCLASS) RETURNS INT
```
Some documentation of the possible status values would be nice to have either here or in the `ts_chunk_status()` definition.

mkindahl replied:
The constants are documented in `chunk.c`, but added a reference to them in the moved function.

mfundul commented on tsl/test/isolation/expected/compression_ddl.out:
```diff
@@ -372,25 +372,28 @@ lock_chunktable
+	FROM _timescaledb_catalog.hypertable ht,
+	     _timescaledb_catalog.chunk ch
+       WHERE ch.hypertable_id = ht.id
+	 AND ht.table_name like 'ts_device_table'
```
Where does this `AND` align? I'm confused with the indentation here.

mkindahl replied:
There is a tab in the specification file, causing this strange indentation. Removed all tabs in the file.

gayyappan reviewed (commented):
Minor test change requested.
Doesn't this need a change for extension updates in in latest-dev.sql and reverse-dev.sql to drop recompress_chunk?

mkindahl commented:
> Minor test change requested.

Don't think it was critical, but added it anyway.

> Doesn't this need a change for extension updates in in latest-dev.sql and reverse-dev.sql to drop recompress_chunk?

They are already added?

gayyappan reviewed (approved):
Missed the latest-dev and reverse-dev changes in the PR.
